src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,835 @@
+                   out_lens.end(),
+                   std::back_inserter(new_dims),
+                   [&](auto len) -> dimension {
+                       return {{dimension::sub{len, {}}}};
+                   });
+    assert(out_lens.size() == new_dims.size());
+    dimensions = new_dims;
+    return true;
+}
+
+// Remove subdimensions of 1
+void remove_1_sub_dims(std::vector<dimension::sub>& subdimensions)
+{
+    subdimensions.erase(std::remove_if(subdimensions.begin(),
+                                       subdimensions.end(),
+                                       [&](const dimension::sub& d) { return d.len == 1; }),
+                        subdimensions.end());
+}
+
+void dimension::simplify()
+{
+    if(subdimensions.size() < 2)
+        return;
+    remove_1_sub_dims(subdimensions);
+    // Flatten adjacent dimensions
+    adjacent_for_each(subdimensions.begin(), subdimensions.end(), [&](sub& d1, sub& d2) {
+        if(d1.axis.size() < 2)
+            return;
+        if(d2.axis.size() < 2)
+            return;
+        if(not std::equal(d1.axis.begin(), d1.axis.end() - 1, d2.axis.begin(), d2.axis.end() - 1))
+            return;
+        auto a1 = d1.axis.back();
+        auto a2 = d2.axis.back();
+        if(a2 <= a1)
+            return;
+        if((a2 - a1) != 1)
+            return;
+        d2.len = d1.len * d2.len;
+        d1.len = 1;
+    });
+    remove_1_sub_dims(subdimensions);
+}
+
+// Search all subdimensions and return the subdimensions vector, an iterator
+// to the subdimension found and an optional iterator to the previous
+// subdimension if available.
+template <class Predicate>
+static auto find_subdimension(shape_transform_descriptor& td, Predicate p)
+{
+    dimension* prev_dim = nullptr;
+    for(auto& d : td.dimensions)
+    {
+        auto it = std::find_if(d.subdimensions.begin(), d.subdimensions.end(), p);
+        if(it != d.subdimensions.end())
+        {
+            decltype(std::make_optional(it)) prev = nullopt;
+            if(it == d.subdimensions.begin())
+            {
+                if(prev_dim != nullptr and not prev_dim->subdimensions.empty())
+                {
+                    prev = std::prev(prev_dim->subdimensions.end());
+                }
+            }
+            else
+            {
+                prev = std::prev(it);
+            }
+            return std::make_tuple(&d.subdimensions, it, prev);
+        }
+        prev_dim = &d;
+    }
+    MIGRAPHX_THROW("Searching for non-existent subdimension");
+}
+
+static bool is_broadcast_dim(const dimension& d)
+{
+    if(d.len() == 1)
+        return false;
+    assert(not d.subdimensions.empty());
+    if(d.subdimensions.size() != 1)
+        return false;
+    const auto& sub = d.subdimensions.front();
+    return sub.axis.empty();
+}
+static bool missing_leading_axis(const dimension& d)
+{
+    if(d.subdimensions.empty())
+        return true;
+    const auto& sub = d.subdimensions.front();
+    return sub.axis.empty();
```

Comment:
Why is it only checking if the first subdimension is empty ? 
Broadcasted subdimension could be anywhere in the list.

Reply:
It using this to group the broadcasted dimensions(using the `group_find` algorithm). So the one-past-the-end should have an axis on the first subdimension so it can figure out where to place the dimensions of 1 which could be broadcasted.

It is true that within the group it may include dimensions that start with a broadcasted subdimension but may include other dimensions so we probably need to filter it out when we construct the `std::deque`.